src/realm/object-store/sync/subscribable.hpp

```diff
@@ -0,0 +1,137 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SUBSCRIBABLE_HPP
+#define REALM_SUBSCRIBABLE_HPP
+
+#include <cstdint>
+#include <unordered_map>
+#include <mutex>
+
+namespace realm {
+
+/// Generic subscribable that allows for coarse, manual notifications
+/// from class type T.
+template <class T>
+struct Subscribable {
+    /// Token that identifies an observer.
+    /// Unsubscribes when deconstructed to
+    /// avoid dangling observers.
+    struct Token {
+        Token(Subscribable* subscribable, uint64_t token)
+        : m_subscribable(subscribable)
+        , m_token(token)
+        {
+        }
+        Token(Token&& other) noexcept
+        : m_subscribable(std::move(other.m_subscribable))
+        , m_token(std::move(other.m_token))
+        {
+            other.m_subscribable = nullptr;
+        }
+        Token& operator=(Token&& other) {
+            m_subscribable = std::move(other.m_subscribable);
+            m_token = std::move(other.m_token);
+            other.m_subscribable = nullptr;
+        }
+        Token(const Token&) = delete;
+        Token& operator=(const Token&) = delete;
+
+        ~Token()
+        {
+            if (m_subscribable) {
+                m_subscribable->unsubscribe(*this);
+            }
+        }
+
+        uint64_t value() const
+        {
+            return m_token;
+        }
+
+    private:
+        Subscribable* m_subscribable;
+        uint64_t m_token;
+
+        template <class U> friend struct Subscribable;
+    };
+
+    using Observer = std::function<void(const T&)>;
+
+    Subscribable() = default;
+    Subscribable(const Subscribable& other)
+    : m_subscribers(other.m_subscribers)
+    {
+    }
+    Subscribable(Subscribable&& other)
+    : m_subscribers(std::move(other.m_subscribers))
+    {
+    }
+    Subscribable& operator=(const Subscribable& other) {
+        m_subscribers = other.m_subscribers;
+    }
+    Subscribable& operator=(Subscribable&& other) {
+        m_subscribers = std::move(other.m_subscribers);
```

Comment:
do we need to hold other's lock here and in the other constructors?

Reply:
For move it makes sense. Does it make sense for copy?

Reply:
I guess the broader question is whether we expect Subscribable to be mutated while it is being assigned to a new value. If yes, then we should hold the lock for both move/copy constructors/assignment-operators, if no then this is fine as-is.

Reply:
It is highly unlikely and would be unexpected, but there's little harm in holding the lock here. I don't expect this to ever be a hot path.